### PR TITLE
feat: wrap vue web app in electron

### LINKED ISSUE #334: v2 - desktop version
**Describe what problem your feature request solves**
We should have a desktop version in place before version 2.0 release

**Describe the solution you'd like**
It would be good to use an electron wrapper for the web application, and this will provide the desktop version of the threat dragon application

**Additional context**
Note that the existing version 1.x threat dragon desktop uses common code with the web application, but does not wrap the web application itself.  This has caused an increase in development and test effort, so it would be good to change the way this is done. Refer to [this example](https://jojozhuang.github.io/tutorial/converting-web-app-to-desktop-app-with-electron/) but it can be done in various ways

### PR DESCRIPTION
**Pull request by ShubhamPalriwala**
Signed-off-by: ShubhamPalriwala <spalriwalau@gmail.com>

**Summary**

Wraps the existing Vue web app using the `electron-builder` vue [plugin](https://nklayman.github.io/vue-cli-plugin-electron-builder/).

**Steps Used:**

```sh
pnpm install vue-cli-plugin-electron-builder
vue add electron-builder
yarn electron:serve (To start dev mode)
```

**Demo:**
[Demo Video](https://streamable.com/e3aq1n)

**Description for the changelog**
Fixes #334

### CODE REVIEW

**Comment by jgadsden** [comment]
Hello @ShubhamPalriwala - looks good and I will checkout the code and run it
Your editor has reformatted the package.json to have indents of 4 spaces, but we want to keep to indents of 2 spaces - could you go back to 2 space indent for this file please? Thanks - and I will have a look at the desktop version

**Comment by ShubhamPalriwala** [comment]
Oh! Sorry for the trouble! I've pushed another commit with the fix in tab size. Lmk if you want me to squash both of them into 1 when you have tested it.

You can test the the feature using:
```
yarn electron:build: To build the app (Production),
yarn electron:serve: To run the app (Development),
```

**Comment by lreading** [comment]
This is looking great, thanks so much for tackling this one @ShubhamPalriwala !

I think it'd be best if we were to remove the need to run `vue add electron-builder`.  I didn't read your clearly written instructions before attempting to run it, and ran into an error. :sweat_smile: I was able to `electron:serve` (without running the vue cli command) by adding `electron-builder` as a dev dependency.  This would just eliminate a step for anyone who tries to build/run locally in the future.  I don't think we would want to utilize a pre-install script or similar, as that might add some overhead to the docker builds.  This is not critical, just my thoughts!  

I tested on Ubuntu, and I'll have a check on Mac/Windows in the next day or two just to confirm.  I wouldn't let that hold this up though, as it's a new feature and we can raise bugs if anything breaks.


Again, _awesome_ work @ShubhamPalriwala!

File: td.vue/package.json
Comment (jgadsden) [inline td.vue/package.json:64]: Could we use the latest version of electron-devtools-installer which is 3.2.0 ?

File: td.vue/package.json
Comment (jgadsden) [inline td.vue/package.json:63]: The latest  version of electron is 17.0.0 - I have tried it with this version and it seems to work OK

File: td.vue/package.json
Comment (jgadsden) [inline td.vue/package.json:80]: latest version of spectron is 19.0.0, can we use this latest version?

File: td.vue/package.json
Comment (jgadsden) [inline td.vue/package.json:27]: if we rename away from the default `background.js` then this line may need to be updated

File: td.vue/src/background.js
Comment (jgadsden) [inline td.vue/src/background.js:1]: can we rename this away from the default `background.js` ?
something with `electron` and `td` in it would indicate what the file is used for ... not worried what the name is so long as it is clear

**Review by jgadsden — COMMENTED** [review]
Hello @ShubhamPalriwala - this is a great first step. I had some comments about the default names / package revisions, if you could have a look at them.

The electron  tests did not  run for me:
```

 FAIL  tests/unit/electron.spec.js (55.769s)
  ✕ Window Loads Properly (50026ms)

  ● Window Loads Properly

    : Timeout - Async callback was not invoked within the 50000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 50000ms timeout specified by jest.setTimeout.Error:

       6 | jest.setTimeout(50000);
       7 |
    >  8 | test('Window Loads Properly', async () => {
         | ^
       9 |     // Wait for dev server to start
      10 |     const { app, stopServe } = await testWithSpectron(spectron);
      11 |     const win = app.browserWindow;

      at new Spec (node_modules/.pnpm/jest-jasmine2@24.9.0/node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Object.<anonymous> (tests/unit/electron.spec.js:8:1)
```

**Review by jgadsden — APPROVED** [review]
I have opened a discussion page for the V2.0 electron feature:
https://github.com/OWASP/threat-dragon/discussions/363

This pull request is a great step towards that feature, so gets my appproval and `i suggest we go ahead and merge 
Requesting review also by @lreading

**Comment by jgadsden** [comment]
I think this is only wrapping td.vue - we also need td.server. I can look into this if you like @ShubhamPalriwala ?

**Comment by lreading** [comment]
> I think this is only wrapping td.vue - we also need td.server. I can look into this if you like @ShubhamPalriwala ? I think others have the same problem : https://stackoverflow.com/questions/69322243/electron-building-issue-for-server-along-with-the-front-end
> 
> when running command `v2-electron/td.vue$ yarn electron:serve` and accessing github we get
> 
> ```
> Proxy error: Could not proxy request /api/login/github from localhost:8080 to http://localhost:3000.
> See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (ECONNREFUSED).
> ```
> 
> If I first run the server from the command line with `v2-electron$ npm run dev:server` then the `v2-electron/td.vue$ yarn electron:serve` can access github as expected

My original plan was to add some wrapping to the auth methods to lock it down to local only when it's running in electron.  I do have a selector in td.vue somewhere to determine if it's running in an electron instance, it's just un-tested and not implemented yet.

We don't _have_ to do it that way though, we could wrap the server in electron as well.  I did some really light research on it a while ago, but haven't ever actually done it.


All that said, that feels like work for another day.  This is a huuuuuuge step forward and great start. <3

**Review by lreading — APPROVED** [review]
Nothing to add here, I think Jon raised some good points as well.  Thanks again for the great work!

**Comment by ShubhamPalriwala** [comment]
> 



> This is looking great, thanks so much for tackling this one @ShubhamPalriwala !
> 
> I think it'd be best if we were to remove the need to run `vue add electron-builder`. I didn't read your clearly written instructions before attempting to run it, and ran into an error. sweat_smile I was able to `electron:serve` (without running the vue cli command) by adding `electron-builder` as a dev dependency. This would just eliminate a step for anyone who tries to build/run locally in the future. I don't think we would want to utilize a pre-install script or similar, as that might add some overhead to the docker builds. This is not critical, just my thoughts!
> 
> I tested on Ubuntu, and I'll have a check on Mac/Windows in the next day or two just to confirm. I wouldn't let that hold this up though, as it's a new feature and we can raise bugs if anything breaks.
> 
> Again, _awesome_ work @ShubhamPalriwala!

I am totally in for anything that makes the user's life easier. I have added the `electron-builder` as a dev dependency and tested the same. Thanks for the quick feedback!

File: td.vue/package.json
Comment (ShubhamPalriwala) [reply td.vue/package.json:64]: Done!

File: td.vue/package.json
Comment (ShubhamPalriwala) [reply td.vue/package.json:27]: Removed this line! Thank you for your suggestion.

**Comment by ShubhamPalriwala** [comment]
> Hello @ShubhamPalriwala - this is a great first step. I had some comments about the default names / package revisions, if you could have a look at them.
> 
> The electron tests did not run for me:
> 
> ```
> 
>  FAIL  tests/unit/electron.spec.js (55.769s)
>   ✕ Window Loads Properly (50026ms)
> 
>   ● Window Loads Properly
> 
>     : Timeout - Async callback was not invoked within the 50000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 50000ms timeout specified by jest.setTimeout.Error:
> 
>        6 | jest.setTimeout(50000);
>        7 |
>     >  8 | test('Window Loads Properly', async () => {
>          | ^
>        9 |     // Wait for dev server to start
>       10 |     const { app, stopServe } = await testWithSpectron(spectron);
>       11 |     const win = app.browserWindow;
> 
>       at new Spec (node_modules/.pnpm/jest-jasmine2@24.9.0/node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
>       at Object.<anonymous> (tests/unit/electron.spec.js:8:1)
> ```

I'll take a look at this and push a fix for the tests. Thank you for pointing this out, really sorry I missed it.

File: td.vue/src/background.js
Comment (ShubhamPalriwala) [reply td.vue/src/background.js:1]: Agreed! How about renaming this to `td.desktop.js`?

Comment (jgadsden) [reply td.vue/src/background.js:1]: Yes, sure - `td.desktop.js` sounds good to me

File: td.vue/package.json
Comment (ShubhamPalriwala) [reply td.vue/package.json:63]: Done!

File: td.vue/package.json
Comment (ShubhamPalriwala) [reply td.vue/package.json:80]: Done! Thank you

File: td.vue/src/background.js
Comment (ShubhamPalriwala) [reply td.vue/src/background.js:1]: Done! Thank you for this suggestion too!

**Comment by ShubhamPalriwala** [comment]
Pushed a commit w all the suggestions incorporated till now. Only thing left is the Electron test issue for which I have opened an issue as I saw a lot of people facing the same issue but none of their solutions worked for us.
Waiting for the response for it [here](https://github.com/nklayman/vue-cli-plugin-electron-builder/issues/1657)

**Comment by ShubhamPalriwala** [comment]
If you are fine with adding the test file later then I can remove the failing test file and then we can merge it!

**Comment by jgadsden** [comment]
All done, thanks again @ShubhamPalriwala

**Comment by ShubhamPalriwala** [comment]
Hey @jgadsden We had to remove the failing test case file :((
Should I open another PR removing that file? 
Because right now the entire tests suite will fail due to this one test case

**Comment by jgadsden** [comment]
No problem - you should have write access? It will make sure we fix it ... I would rather fix it than remove it? I can look at getting the tests running this morning

**Comment by ShubhamPalriwala** [comment]
Thank you so much! I'm waiting for the update here https://github.com/nklayman/vue-cli-plugin-electron-builder/issues/1657.

See if you can do anything!

**Pull request merged**
